accepted/0000-subscription-matching-in-public-clouds.md

```diff
@@ -0,0 +1,194 @@
+for i in response['Reservations']:
+    for j in i['Instances']:
+        print(j)
+```
+
+The response contains the "instance name" as well as the "instance id".
+
+### Google Compute Engine module:
+
+This is an example about how to deal with Google Compute Engine API to collect the virtual instances on a particular Zone, using "google-api-python-client":
+
+```python
+import googleapiclient.discovery
+
+def list_instances(compute, project, zone):
+    result = compute.instances().list(project=project, zone=zone).execute()
+    return result['items'] if 'items' in result else None
+
+    zone = 'MY-ZONE'
+    project = 'MY-PROJECT-NAME'
+
+    compute = googleapiclient.discovery.build('compute', 'v1')
+
+instances = list_instances(compute, project, zone)
+for i in instances:
+    print(i)
+```
+
+The response contains the "instance name" as well as the "instance id".
+
+### A generic JSON-file module:
+
+Currently, there is already a "File based" plugin for `virtual-host-gatherer` which allows to import instances from a custom provided JSON file. According to the existing documentation, the aim of this plugin is to import VMware instances when there is no access from the SUSE Manager server to the VMware.
+
+We can already use this plugin for importing virtual instances from the Public Cloud using a tailored JSON file like the one from the above example, so it would be worth to generalize the SUSE Manager documentation to expose this plugin as a general virtual instances importer and not as a VMware-specific.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Gathering the UUID (instance id) from virtual instances
+
+For AWS and GCE, the instance id that is returned from the API is not actually the real "uuid" from the instance. On Azure, the instance id is an UUID but not necessary correspond with the SMBIOS "uuid" value we get from Salt.
+
+An easy approach here would be to use the "Instance ID" (instead of smbios uuid) when registering a system which is a public cloud virtual instance. Salt currently does not provide the instance id as part of the grains but it would be really easy to provide a custom grain at the time of registration that would expose the "instance id" as part of the grains only when the system is an EC2, GCE or Azure instance. [Example here](https://gist.github.com/meaksh/1ed58ece0f26ce27a8445985de9ad6a2)
+
+This way, doing some minor fixes on the Java side ([example here](https://github.com/meaksh/uyuni/commit/03d88550dd87d22f3fabd25cebd7c23432285a3c)), we could easily use the "instance-id" as "UUID" for the registered system and automatically match it with the data provided by the `virtual-host-gatherer` plugin (which does not include "uuid" but instance id).
+
+In case of systems that are already registered in SUSE Manager using a smbios "uuid", if the new "instance_id" grain is there, it should be enough with scheduling a "Hardware Refresh" action to reflect the new "instance_id" grain value as the "uuid" for that system.
```

Comment:
Please put this in the "Detailed design" section. It is exactly what I had in mind. It is not a drawback, it the way how we should implement it.

Another question: what about "traditional" clients? Should we implement that as well, or can we say that this is only supported with salt (@paususe) ?

Traditional  clients and salt is both python. So we could use the same code. But we would need to enhance the traditional registration protocol and the server side of that is in python, not in java.

@meaksh that could be a "unresolved question" and we can discuss it there.

Reply:
It's in our best interest as a company to detect as many subscriptions as possible.

How difficult is the traditional client case, in addition to salt? I'd do salt clients first, then traditional.

Reply:
I guess we would need to do some changes on the Traditional Client code (registration & hardware refresh) to provide the "uuid" based on the "instance id" (instead of smbios) if we detect the instance is running in a Public Cloud environment. So introducing the same code that we have as example of a custom "instance_id" grain on Salt.